File: eclipse_plugin/src/com/sina/weibo/agent/eclipse/util/NodeVersionUtil.java
/*
 * SPDX-FileCopyrightText: 2025 Weibo, Inc.
 * SPDX-License-Identifier: Apache-2.0
 */
package com.sina.weibo.agent.eclipse.util;

import java.io.BufferedReader;
import java.io.InputStreamReader;

import com.sina.weibo.agent.eclipse.Activator;

/**
 * Utility class for Node.js version checking.
 */
public class NodeVersionUtil {

    /**
     * Check if Node.js version satisfies the minimum requirement.
     *
     * @param nodePath the path to Node.js executable
     * @param minVersion the minimum required version (e.g., "18.0.0")
     * @return true if version is satisfied
     */
    public static boolean isVersionSatisfied(String nodePath, String minVersion) {
        String currentVersion = getNodeVersion(nodePath);
        if (currentVersion == null) {
            return false;
        }
        return compareVersions(currentVersion, minVersion) >= 0;
    }

    /**
     * Get Node.js version.
     *
     * @param nodePath the path to Node.js executable
     * @return the version string or null
     */
    public static String getNodeVersion(String nodePath) {
        try {
            ProcessBuilder builder = new ProcessBuilder(nodePath, "--version");
            builder.redirectErrorStream(true);
            Process process = builder.start();

            try (BufferedReader reader = new BufferedReader(
                    new InputStreamReader(process.getInputStream()))) {
                String line = reader.readLine();
                if (line != null && line.startsWith("v")) {
                    return line.substring(1).trim();
                }
            }

            process.waitFor();
        } catch (Exception e) {
            Activator.logError("Failed to get Node.js version", e);
        }
        return null;
    }

    /**
     * Compare two version strings.
     *
     * @param version1 first version
     * @param version2 second version
     * @return negative if version1 < version2, zero if equal, positive if version1 > version2
     */
    public static int compareVersions(String version1, String version2) {
        String[] parts1 = version1.split("\\.");
        String[] parts2 = version2.split("\\.");

        int maxLength = Math.max(parts1.length, parts2.length);

        for (int i = 0; i < maxLength; i++) {
            int v1 = i < parts1.length ? parseVersionPart(parts1[i]) : 0;
            int v2 = i < parts2.length ? parseVersionPart(parts2[i]) : 0;

            if (v1 != v2) {
                return v1 - v2;
            }
        }

        return 0;
    }

    /**
     * Parse a version part (handle non-numeric suffixes).
     *
     * @param part the version part
     * @return the numeric value
     */
    private static int parseVersionPart(String part) {
        // Remove any non-numeric suffix (e.g., "18.12.0-rc.1" -> "18")
        StringBuilder numStr = new StringBuilder();
        for (char c : part.toCharArray()) {
            if (Character.isDigit(c)) {
                numStr.append(c);
            } else {
                break;
            }
        }
        
        if (numStr.length() == 0) {
            return 0;
        }
        
        try {
            return Integer.parseInt(numStr.toString());
        } catch (NumberFormatException e) {
            return 0;
        }
    }
}

File: eclipse_plugin_new/src/main/kotlin/com/sina/weibo/agent/util/NodeVersionUtil.kt
// SPDX-FileCopyrightText: 2025 Weibo, Inc.
//
// SPDX-License-Identifier: Apache-2.0

package com.sina.weibo.agent.util

import org.eclipse.core.runtime.Platform
import org.osgi.framework.FrameworkUtil
import java.io.BufferedReader
import java.io.File
import java.io.InputStreamReader

/**
 * Node version utility class
 * Get local Node.js version and perform version checks
 */
object NodeVersionUtil {
    private val LOG = Platform.getLog(FrameworkUtil.getBundle(NodeVersionUtil::class.java))

    /**
     * Get local Node.js version
     *
     * @return Node.js version string, or null if not found
     */
    fun getLocalNodeVersion(): String? {
        val nodeExecutable = findNodeExecutable()
        if (nodeExecutable == null) {
            LOG.warn("Node.js executable not found")
            return null
        }

        return try {
            val process = ProcessBuilder(nodeExecutable.absolutePath, "--version").start()
            val reader = BufferedReader(InputStreamReader(process.inputStream))
            val version = reader.readLine()
            process.waitFor()
            version
        } catch (e: Exception) {
            LOG.error("Failed to get Node.js version", e)
            null
        }
    }

    /**
     * Find Node.js executable
     * Search in common installation locations and PATH environment variable
     */
    private fun findNodeExecutable(): File? {
        // Search in PATH environment variable
        val pathDirs = System.getenv("PATH").split(File.pathSeparator)
        for (dir in pathDirs) {
            val nodeFile = File(dir, if (System.getProperty("os.name").startsWith("Windows")) "node.exe" else "node")
            if (nodeFile.exists() && nodeFile.canExecute()) {
                return nodeFile
            }
        }

        // Search in common installation locations
        val commonPaths = arrayOf(
            "/usr/local/bin/node",
            "/usr/bin/node",
            "C:\\Program Files\\nodejs\\node.exe"
        )
        for (path in commonPaths) {
            val nodeFile = File(path)
            if (nodeFile.exists() && nodeFile.canExecute()) {
                return nodeFile
            }
        }

        return null
    }

    /**
     * Check if Node.js version meets minimum requirement
     *
     * @param requiredVersion Minimum required version (e.g., "14.0.0")
     * @return True if version requirement is met, false otherwise
     */
    fun isNodeVersionSufficient(requiredVersion: String): Boolean {
        val localVersion = getLocalNodeVersion()
        if (localVersion == null) {
            LOG.warn("Cannot check Node.js version because it is not installed or not found")
            return false
        }

        return try {
            val local = Version(localVersion)
            val required = Version(requiredVersion)
            local >= required
        } catch (e: Exception) {
            LOG.error("Failed to compare Node.js versions", e)
            false
        }
    }

    /**
     * Simple version comparison class
     */
    private data class Version(val version: String) : Comparable<Version> {
        private val parts = version.removePrefix("v").split(".").map { it.toInt() }

        override fun compareTo(other: Version): Int {
            val size = maxOf(parts.size, other.parts.size)
            for (i in 0 until size) {
                val p1 = parts.getOrElse(i) { 0 }
                val p2 = other.parts.getOrElse(i) { 0 }
                if (p1 != p2) {
                    return p1 - p2
                }
            }
            return 0
        }
    }
}
